db/functions/base.py

```diff
@@ -13,15 +13,46 @@
 """
 
 from abc import ABC, abstractmethod
+import warnings
 
-from sqlalchemy import column, not_, and_, or_, func, literal, cast, INTEGER
-from sqlalchemy.dialects.postgresql import array_agg
+from sqlalchemy import column, not_, and_, or_, func, literal
+from sqlalchemy.dialects.postgresql import array_agg, INTEGER, TEXT
+from sqlalchemy.sql import quoted_name
+from sqlalchemy.sql.functions import GenericFunction, concat
 
 from db.functions import hints
 from db.functions.exceptions import BadDBFunctionFormat
 
 
-def sa_call_sql_function(function_name, *parameters):
+def sa_call_sql_function(function_name, *parameters, return_type=None):
+    """
+    This function registers an SQL function with SQLAlchemy and generates
+    an expression to call it.
+
+    function_name: string giving a namespaced SQL function
+    *parameters:   these will be passed directly to the generated function.
+    return_type:   an SQLAlchemy type class
+    """
+    if return_type is None:
+        warnings.warn(
+            "sa_call_sql_function should be called with the return_type kwarg set"
+        )
+        # We can't use PostgresType since we don't want an engine here
+        return_type = TEXT
+
+    with warnings.catch_warnings():
+        warnings.filterwarnings("ignore", message="The GenericFunction")
+        # Creating this type registers the function (more importantly,
+        # its return type) with SQLAlchemy. **magic!!**
+        type(
```

Comment:
I wouldn't expect this `type` call to have side-effects. Shouldn't it be assigned to something?

Reply:
That's why it's magic. It registers a type with the python interpreter, which gets picked up by SQLAlchemy.

I also find it arcane, hence the comment. Maybe I can extend that comment it a bit?

Reply:
Oh, that's what the comment above talks about. Magic indeed.

Reply:
Now that I get it, the comment makes sense, but that took me a while. Might want to add a few more details or something. Not requesting this change though.